### ISSUE TITLE
Unable to trigger Conditional Access flow

### ISSUE THREAD

**Issue by mlynch** [msintuneappsdk/ms-intune-app-sdk-android#87]
**Describe the bug:**

I am trying to test a Conditional Access auth flow using Intune and MSAL. On iOS I am able to successfully trigger the Conditional Access flow. On Android, I never receive a `MsalIntuneAppProtectionPolicyRequiredException` event in the `onError` of my `AuthenticationCallback` so I never have a chance to call `remediateCompliance` on the `MAMComplianceManager`.

I have checked to make sure I have `"client_capabilities": "protapp",` in my `auth_config.json`, as well as calling `forceRefresh(false)` on the params for `acquireTokenSilentAsync`.

The issue is I'm not seeing any errors. My auth flow works fine and I receive a token both in the silent and non-silent MSAL flows. When I look at the sign-in logs I see under `Conditional Access` a `Success` value, but when I click on the sign-in entry and go to the `Conditional Access` tab I see a Failure under `Grant Controls` which is very confusing.

On iOS I do get the compliance flow to trigger and auth only works when I have it correctly set up, so I'm not really sure what to do.

**To Reproduce**

**Expected behavior:**

Expect the compliance errors to fire when requesting a token so I can engage the `MAMComplianceManager` flow per the Android App SDK docs

**Screenshots and logs:**

My Conditional Access policy:

<img width="935" alt="Screen Shot 2021-12-03 at 5 41 56 PM" src="https://user-images.githubusercontent.com/11214/144686756-16a4eb8a-a9c2-47d6-9e9e-bf6e8505e3e2.png">

<img width="1443" alt="Screen Shot 2021-12-03 at 5 46 58 PM" src="https://user-images.githubusercontent.com/11214/144686907-2d01a979-14d1-4add-a079-0b8f0a5bc973.png">
<img width="1411" alt="Screen Shot 2021-12-03 at 5 48 08 PM" src="https://user-images.githubusercontent.com/11214/144686908-cc849fec-3809-4feb-8b46-7480cab0158a.png">

**Smartphone (please complete the following information):**
 - Device: Samsung Galaxy A10
 - OS: Android 9

**Intune App SDK for Android (please complete the following information):**
- What version of the Intune SDK are you using? Are you using the latest version? 8.1.1
- What platform is your app based in (Java, Xamarin based, Cordova, etc)? [Capacitor](https://capacitorjs.com/)

**Comment by mlynch** [comment]
I talked with our client who is struggling to integrate Intune in their app. Their conditional access flow appears to be engaging and breaking their login, but when we try to duplicate it on our end we are not able to get it to trigger and auth works correctly. Any help on the above would be greatly appreciated, as right now we're both stuck.

**Comment by bannus** [comment]
@mlynch Which version of MSAL are you using? Are you able to either share a sample app that can reproduce the issue or expand on your repro steps? (Are those steps supposed to say '...' repeatedly?)

**Comment by mlynch** [comment]
Ah apologies didn't mean to leave those in. I'm still trying to get more info from the customer. It sounds like they *might* be having a different issue.

I'm still not able to get the compliance manager/conditional access flow to actually trigger though. Is there any example of this working? With the same configuration it triggers on iOS so I'm very confusion

**Comment by Ddv0623** [comment]
Hi,
I have same issue.
When get token on iOS there will be MsalIntuneAppProtectionPolicyRequiredException then in code I will remediateCompliance.
For Android it get token without any issue or exception. Is it expected?

I using MSAL 2.2.2

**Comment by mlynch** [comment]
@Ddv0623 glad I'm not the only one! Unfortunately, this API is very obscure and there don't appear to be any examples of how to use this API

**Comment by Ddv0623** [comment]
@bannus Do Microsoft have implement a full example how to implement Intune with CA support for android using java. We can take a look that example to check if we missing any thing.
Because so far when I get token by MSAL there are no error return where there are should be an error because the app is not protected at that time.

**Comment by joakley-msft** [comment]
@Ddv0623 our sample app at https://github.com/msintuneappsdk/Taskr-Sample-Intune-Android-App/ does demonstrate some MAM CA capabilities. Let us know if this helps at all.

**Comment by mlynch** [comment]
Thanks @joakley-msft, all I can find is this exception handler here, but it seems like it's not implementing the CA stuff beyond just a note on how to do it: https://github.com/msintuneappsdk/Taskr-Sample-Intune-Android-App/blob/ca9fa2f9b350ee0c3a8566cc4ac5dad4ad9fbbfc/app/src/main/java/com/microsoft/intune/samples/taskr/MainActivity.java#L204

Unfortunately, it seems like @Ddv0623 and I can't get the CA stuff to trigger at all in practice on Android.

**Comment by joakley-msft** [comment]
We're tracking the sample improvement internally with #13121189.

When you say you can't get the CA stuff to trigger at all -- does that mean that your calls to MSAL succeed and give you an access token, despite no MAM policy being applied to the app?

**Comment by mlynch** [comment]
Correct, we get an access token back, but we also do get a working Intune MAM policy on Android and no Conditional Access errors raised. It's as-if the CA stuff is just doing nothing on Android.

However, on iOS we *do* get the CA errors raised and have to call `IntuneMAMComplianceManager.instance().remediateCompliance`. The configuration on the Azure side looks correct and identical for both iOS and Android.

**Comment by joakley-msft** [comment]
The part that has me a bit puzzled here is that by definition Conditional Access is enforced service-side in AAD. In the case of MAM-based CA, you're conditionally given an access token only if policy is applied to your app. There isn't anything you can do "wrong" in your app to accidentally get an access token without Intune thinking it's applying MAM policy. And that fact that your iOS app is getting blocked by CA for what I presume is the same resource does seem to indicate that your CA policies are configured correctly.

At the time you're signing in to your app on Android, is the Company Portal installed, or not yet? I'm wondering if it could be that policy is *already* applied.

**Comment by joakley-msft** [comment]
@mlynch in that case could you follow the directions at https://docs.microsoft.com/en-us/mem/intune/user-help/send-logs-to-microsoft-android to upload logs. Please relay the incident id here and we can take a look.

**Comment by sandeepbhatia4922** [comment]
@mlynch  exception MsalIntuneAppProtectionPolicyRequiredException is thrown and handled inside on error() method. But MAM policies not enforced using 
`val complianceManager = MAMComponents.get(MAMComplianceManager::class.java)
                complianceManager?.remediateCompliance(upn, aaid, tenantis, authorityURL, true)`
                
The reason is I am getting aaid null from exception.

`if (exception is MsalIntuneAppProtectionPolicyRequiredException) {

                // Note: An app that has enabled APP CA with Policy Assurance would need to pass these values to `remediateCompliance`.
                // For more information, see https://docs.microsoft.com/en-us/mem/intune/developer/app-sdk-android#app-ca-with-policy-assurance
                val upn = exception.accountUpn
                val aadid = exception.accountUserId // returns always null
                val tenantId = exception.tenantId
                val authorityURL = exception.authorityUrl`

**Comment by joakley-msft** [comment]
> I have raised same issue here https://github.com/msintuneappsdk/ms-intune-app-sdk-android/issues/106

Closing this issue in favor of #106 so we're not discussing the same thing in two places.
